src/backend/src/modules/captcha/middleware/captcha-middleware.js

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2024-present Puter Technologies Inc.
+ *
+ * This file is part of Puter.
+ *
+ * Puter is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const APIError = require("../../../api/APIError");
+const { Context } = require("../../../util/context");
+
+/**
+ * Middleware that requires captcha verification
+ * Can be used in two modes: always require or conditionally require
+ * 
+ * @param {Object} options - Configuration options
+ * @param {boolean} [options.always=false] - Always require captcha regardless of rate limits
+ * @param {boolean} [options.strictMode=true] - If true, fails closed on errors (more secure)
+ * @returns {Function} Express middleware function
+ */
+const requireCaptcha = (options = {}) => async (req, res, next) => {
+    // Default to strict mode for security
+    const strictMode = options.strictMode !== false;
+    
+    try {
+        // Get services from the Context instead of req.app
+        const services = Context.get('services');
+        
+        // Fail closed if services aren't available (in strict mode)
+        if (!services) {
+            console.warn('Captcha middleware: services not available in Context');
+            if (strictMode) {
+                // Use next(error) instead of throwing directly
+                return next(APIError.create('internal_error', null, {
+                    message: 'Captcha service unavailable',
+                    status: 500
```

Comment:
503 might be more appropriate
```suggestion
                    status: 503
```